website/content/docs/upgrading/instructions/testing-experimental-raft-backend.mdx

```diff
@@ -0,0 +1,420 @@
+  Consul 1.15 introduced a new experimental storage backend option. Learn how to
+  configure and test it out in a safe way.
+---
+
+# Testing the Experimental WAL LogStore Backend
+
+## Introduction
+
+Consul 1.15 introduced a new experimental storage backend. This guide explains
+how to configure it and test it out safely. The new backend is called `wal` in
+configuration.
+
+WAL is a acronym for "Write-Ahead Log". We called it this because it implements
+a traditional log with rotating, append-only log files. The current `LogStore`
+uses BoltDB which is a copy-on-write BTree which is less optimized for
+append-only workloads.
+
+~> The `wal` backend is considered **experimental** in Consul 1.15. Please test
+it safely in pre-production first and verify on a subset of servers in
+production using this guide before fully enabling it on all servers.
+
+## Why build a new backend?
+
+The WAL backend has been written to resolve some long-standing issues with the
+current BoltDB backend. The existing BoltDB log store has worked reliably for
+most users for years, however it is not the most efficient way to store
+append-only logs to disk since it was designed as a full key-value database. It
+was an expedient option when our raft library was first written and always
+assumed we'd replace it with something more purpose-built.
+
+Importantly, a BoltDB database is a single file that only ever grows. Deleting
+the oldest logs which we do regularly when we've made a new snapshots of the
+state, leaves free space in the file that needs to be tracked to be re-used on
+future writes. By contrast a simple segmented log can just delete the oldest log
+files from disk. With BoltDB uses as a log, sudden burst of writes say three
+times larger than the normal volume can suddenly cause the file to grow to about
+several times it's steady-state size. After the next snapshot is taken, and the
+oldest logs truncated again, the file is left as mostly empty space. Tracking
+this free space requires writing extra metadata proportional to the amount of
+free pages to disk with every write and so after such a burst, write latencies
+tend to increase - in some cases dramatically causing serious performance
+degradation to the cluster.
+
+Even if this has never happened to a catastrophic degree in a cluster, the fact
+that it's a risk has meant that Consul has erred on the side of never letting
+too many logs accumulate in the LogStore. Significantly larger BoltDB files are
+somewhat slower in general because it's a tree and so still has log(N) work to
+do n every write. But out user's experience showed that the larger the file, the
+more likely it is to have a large freelist or suddenly form one after a burst of
+writes. For this reason, the default options for how frequently we make a full
+snapshot and truncate the logs, and for how many logs we keep around have always
+been aggressively set towards keeping BoltDB small rather than using disk IO the
+most efficiently.
```

Comment:
```suggestion
Even if this has never happened to a catastrophic degree in a cluster, Consul was tuned to avoid for too many logs to accumulate in the LogStore, in order to reduce/mitigate the risk. Significantly larger BoltDB files are
somewhat slower in general because it's a tree and so still has log(N) work to
do n every write. But out user's experience showed that the larger the file, the
more likely it is to have a large freelist or suddenly form one after a burst of
writes. For this reason, the default options for how frequently we make a full
snapshot and truncate the logs, and for how many logs we keep around have always
been aggressively set towards keeping BoltDB small rather than using disk IO the
most efficiently.
```

Reply:
Also:

> Significantly larger BoltDB files are
somewhat slower in general because it's a tree and so still has log(N) work to
do n every write. But out user's experience showed that the larger the file, the
more likely it is to have a large freelist or suddenly form one after a burst of
writes.

Do we mean that both (the fact it is a tree and the higher probability of freelist) are a reason of bad performances? Or that despite the good performances offered by the tree, freelists drags them down?

> because it's a tree and so still has log(N) work to do n every write
Did we mean `on every write`?

Reply:
Also I am assuming that by `freelist` we refer either to chunks of free space in the file or to the data structures we use to track the free space in the file.

It is not mentioned anywhere else in the file so it might be worth a clarification.